File: acwing/算法提高课/第一章动态规划/1057.股票买卖IV/1057.股票买卖IV.py
N = int(1e5) + 10
K = 110

f = [[float("-inf") for _ in range(2)] for _ in range(K)]

n, k = map(int, input().split())
a = [0] + list(map(int, input().split()))


f[0][0] = 0

for i in range(1, n + 1):
    for j in range(k, -1, -1):
        f[j][0] = max(f[j][0], f[j][1] + a[i])#  f[j][0]和 f[j][1]的计算顺序不能反过来
        f[j][1] = max(f[j][1], f[j - 1][0] - a[i])

res = 0
for i in range(k + 1):
    res = max(res, f[i][0])
print(res)

File: acwing/算法提高课/第一章动态规划/1057.股票买卖IV/1057.股票买卖IV.cpp
#include <algorithm>
#include <cstring>
#include <iostream>

using namespace std;

const int N = 1e5 + 10, K = 110;

int f[K][2];
int a[N];

int main()
{
    int n, k;
    cin >> n >> k;
    for (int i = 1; i <= n; i++)
        cin >> a[i];
    memset(f, -0x3f, sizeof(f));
    f[0][0] = 0;
    for (int i = 1; i <= n; i++)
    {
        for (int j = k; j >= 1; j--)
        {
            f[j][0] = max(f[j][0], f[j][1] + a[i]);
            f[j][1] = max(f[j][1], f[j - 1][0] - a[i]);
        }
    }
    int res = 0;
    for (int i = 0; i <= k; i++)
        res = max(res, f[i][0]);
    cout << res << endl;
    return 0;
}